### PR TITLE
Improve entries apis: Context Pad, Palette, Popup Menu

### LINKED ISSUE #428: Improve entries providers for palette, context pad and popup menu
__Is your feature request related to a problem? Please describe.__

Currently, the palette and context pad providers are not capable of removing entries added by other providers. For popup menu, there can be only one provider for specific id. This limits the extensibility options and usually forces to subclass the extended library's provider.

__Describe the solution you'd like__

* [x] Create a unified and powerful API for different types of providers:
  * [x] provider can add, update and remove entries
  * [x] provider can register itself with a priority
  * [x] provider is called via its `get[componentName][type]Entries` method and can either return the entries to be added (legacy behaviour) or a function that gets called with `entries` and returns the updated `entries`
* [x] Implement the feature for:
  * [x] context pad
  * [x] palette
  * [x] popup menu
* [x] Update JSDocs and examples:
  * [x] context pad
  * [x] palette
  * [x] popup menu

__Describe alternatives you've considered__

Leave it as it is.

__Additional context__

Related to https://github.com/camunda/camunda-modeler/issues/1486

### PR DESCRIPTION
**Pull request by barmac**
Closes #428

Note: This breaks multiple test cases in `bpmn-js`, but all of the failures base on checking the internals. I'll prepare a fix for that.

### CODE REVIEW

**Comment by barmac** [comment]
This can be tested with https://github.com/barmac/bpmn-js-reworked-replace-menu

File: lib/features/context-pad/ContextPad.js [hunk 1]
```diff
@@ -106,10 +108,30 @@ ContextPad.prototype._init = function() {
...
  * Register a provider with the context pad
  *
+ * @param  {Number} [priority=1000]
  * @param  {ContextPadProvider} provider
+ *
+ * @example
+ * contextPad.registerProvider(800, {
+  *   getContextPadEntries: function(element) {
+  *     return function(entries) {
+  *       return Object.assign(entries, {
```
Comment (nikku) [inline lib/features/context-pad/ContextPad.js]: This example is dangerous, as it patches `entries` in place. 

Please use ES6 spread operators here:

```
return {
  ...entries,
  'entry-1': { ... }
};
```

File: lib/features/context-pad/ContextPad.js [hunk 2]
```diff
@@ -106,10 +108,30 @@ ContextPad.prototype._init = function() {
...
  */
-ContextPad.prototype.registerProvider = function(provider) {
-  this._providers.push(provider);
+ContextPad.prototype.registerProvider = function(priority, provider) {
+  if (!provider) {
+    provider = priority;
+    priority = DEFAULT_PRIORITY;
+  }
+
+  this._insertProviderByPriority(provider, priority);
```
Comment (nikku) [inline lib/features/context-pad/ContextPad.js]: I'd like to use the eventBus here, as it handles priorities already. Will prepare a proposal regarding that.

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 3]
```diff
@@ -260,37 +304,14 @@ describe('features/popup', function() {
...
-
-
     it('should throw error when no provider', inject(function(popupMenu) {
 
       // when not registering a provider
 
       // then
       expect(function() {
         popupMenu.open({}, 'foo', { x: 100, y: 100 });
-      }).to.throw('Provider is not registered: foo');
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js]: Why don't we assert the error message anymore?

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 4]
```diff
@@ -342,14 +363,12 @@ describe('features/popup', function() {
 
         // then
         expect(popupMenu._current).to.exist;
-        expect(popupMenu._current.headerEntries[0].id).to.eql('entryA');
-        expect(popupMenu._current.entries[0].id).to.eql('entryB');
+        expect(popupMenu._current.headerEntries['entryA']).to.exist;
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js]: to.have.keys can be a good matcher to express this:

```
expect(popupMenu._current.entries).to.have.keys('entryA', 'entryB');
```

File: lib/features/context-pad/ContextPad.js (hunk 1 above)
Comment (barmac) [reply lib/features/context-pad/ContextPad.js]: OK, I'll update both `ContextPad` and `Palette` examples.

File: test/spec/features/popup-menu/PopupMenuSpec.js (hunk 4 above)
Comment (barmac) [reply test/spec/features/popup-menu/PopupMenuSpec.js]: Let's do that.

File: test/spec/features/popup-menu/PopupMenuSpec.js (hunk 3 above)
Comment (barmac) [reply test/spec/features/popup-menu/PopupMenuSpec.js]: No specific reason. I'll update it.

File: lib/features/context-pad/ContextPad.js (hunk 2 above)
Comment (barmac) [reply lib/features/context-pad/ContextPad.js]: Makes sense, I should've thought about this.

**Comment by barmac** [comment]
@nikku I updated the PR in accordance with the comments.

**Comment by nikku** [comment]
I've sketched how to register providers using the eventBus [here](https://github.com/bpmn-io/diagram-js/pull/431/commits/59681ee6ffc0398ff221130e3d34caa495017752).

File: lib/features/context-pad/ContextPad.js [hunk 5]
```diff
@@ -106,10 +108,31 @@ ContextPad.prototype._init = function() {
...
```
Comment (nikku) [inline lib/features/context-pad/ContextPad.js]: TODO: register via EventBus, cf. https://github.com/bpmn-io/diagram-js/pull/431/commits/59681ee6ffc0398ff221130e3d34caa495017752.

File: lib/features/popup-menu/PopupMenu.js [hunk 6]
```diff
@@ -62,26 +66,32 @@ PopupMenu.$inject = [
  * Registers a popup menu provider
  *
  * @param  {String} id
+ * @param {Number} [priority=1000]
  * @param  {Object} provider
  *
  * @example
  *
  * popupMenu.registerProvider('myMenuID', {
```
Comment (nikku) [inline lib/features/popup-menu/PopupMenu.js]: Maybe separate creating the provider from registering? That would make the example easier to comprehend in my opinion. Cf. https://github.com/bpmn-io/diagram-js/pull/431/commits/59681ee6ffc0398ff221130e3d34caa495017752

File: lib/features/popup-menu/PopupMenu.js [hunk 7]
```diff
@@ -62,26 +66,32 @@ PopupMenu.$inject = [
...
-};
+PopupMenu.prototype.registerProvider = function(id, priority, provider) {
+  var providers = this._providers[id] = this._providers[id] || [];
+
+  if (!provider) {
+    provider = priority;
+    priority = DEFAULT_PRIORITY;
+  }
 
+  this._insertProviderByPriority(providers, provider, priority);
```
Comment (nikku) [inline lib/features/popup-menu/PopupMenu.js]: Register via eventBus instead? Cf. https://github.com/bpmn-io/diagram-js/pull/431/commits/59681ee6ffc0398ff221130e3d34caa495017752

**Comment by barmac** [comment]
@nikku Ready for another round.

**Pull request merged**